modules/javafx.graphics/src/main/native-glass/gtk/glass_window_ime.cpp

```diff
@@ -23,266 +23,157 @@
+                slen,
+                slen,
+                NULL);
+        LOG_EXCEPTION(mainEnv)
+    } else {
+        im_ctx.send_keypress = true;
     }
 }
 
-//Note: this function must return int, despite the fact it doesn't conform to XIMProc type.
-// This is required in documentation of XIM
-static int im_preedit_start(XIM im_xim, XPointer client, XPointer call) {
-    (void)im_xim;
-    (void)call;
-
-    mainEnv->CallVoidMethod((jobject) client, jViewNotifyPreeditMode, JNI_TRUE);
-    CHECK_JNI_EXCEPTION_RET(mainEnv, -1);
-    return -1; // No restrictions
+bool WindowContextBase::hasIME() {
+    return im_ctx.enabled;
 }
 
-static void im_preedit_done(XIM im_xim, XPointer client, XPointer call) {
-    (void)im_xim;
-    (void)call;
-
-    mainEnv->CallVoidMethod((jobject) client, jViewNotifyPreeditMode, JNI_FALSE);
-    CHECK_JNI_EXCEPTION(mainEnv);
-}
+bool WindowContextBase::filterIME(GdkEvent *event) {
+    if (!hasIME()) {
+        return false;
+    }
 
-static void im_preedit_draw(XIM im_xim, XPointer client, XPointer call) {
-    (void)im_xim;
-    (void)call;
-
-    XIMPreeditDrawCallbackStruct *data = (XIMPreeditDrawCallbackStruct*) call;
-    jstring text = NULL;
-    jbyteArray attr = NULL;
-
-    if (data->text != NULL) {
-        if (data->text->string.multi_byte) {
-            if (data->text->encoding_is_wchar) {
-                size_t csize = wcstombs(NULL, data->text->string.wide_char, 0);
-                char *ctext = new char[csize + 1];
-                wcstombs(ctext, data->text->string.wide_char, csize + 1);
-                text = mainEnv->NewStringUTF(ctext);
-                delete[] ctext;
-                CHECK_JNI_EXCEPTION(mainEnv);
-            } else {
-                text = mainEnv->NewStringUTF(data->text->string.multi_byte);
-                CHECK_JNI_EXCEPTION(mainEnv);
-            }
-        }
+    bool filtered = gtk_im_context_filter_keypress(im_ctx.ctx, &event->key);
 
-        if (XIMFeedback* fb = data->text->feedback) {
-            attr = mainEnv->NewByteArray(data->text->length);
-            CHECK_JNI_EXCEPTION(mainEnv)
-            jbyte v[data->text->length];
-            for (int i = 0; i < data->text->length; i++) {
-                if (fb[i] & XIMReverse) {
-                    v[i] = com_sun_glass_ui_View_IME_ATTR_TARGET_NOTCONVERTED;
-                } else if (fb[i] & XIMHighlight) {
-                    v[i] = com_sun_glass_ui_View_IME_ATTR_TARGET_CONVERTED;
-                } else if (fb[i] & XIMUnderline) {
-                    v[i] = com_sun_glass_ui_View_IME_ATTR_CONVERTED;
-                } else {
-                    v[i] = com_sun_glass_ui_View_IME_ATTR_INPUT;
-                }
-            }
-            mainEnv->SetByteArrayRegion(attr, 0, data->text->length, v);
-            CHECK_JNI_EXCEPTION(mainEnv)
-        }
+    if (!filtered || (filtered && im_ctx.send_keypress)) {
+        process_key(&event->key);
+        im_ctx.send_keypress = false;
```

Comment:
I'm seeing two RELEASE events on each keystroke. If you call process_key() here you need to set `filtered` to true to ensure the event isn't processed again.

Reply:
I changed to "if not filtered", just propagate.

Reply:
Looks good to me.